File: tool/unifmu/resources/backends/python/backend.py
import logging
import os
import sys
from schemas.unifmu_fmi2_pb2 import (
    Fmi2Command,
    Fmi2Return,
    Fmi2ExtHandshakeReturn,
    Fmi2ExtSerializeSlaveReturn,
    Fmi2StatusReturn,
)

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__file__)

try:
    import zmq
except ImportError:
    logger.fatal(
        "unable to import the python library 'zmq' required by the schemaless backend. "
        "please ensure that the library is present in the python environment launching the script. "
        "the missing dependencies can be installed using 'python -m pip install unifmu[python-backend]'"
    )
    sys.exit(-1)

from model import Model

if __name__ == "__main__":

    slave = Model()

    # initializing message queue
    context = zmq.Context()
    socket = context.socket(zmq.REQ)
    dispatcher_endpoint = os.environ["UNIFMU_DISPATCHER_ENDPOINT"]
    logger.info(f"dispatcher endpoint received: {dispatcher_endpoint}")
    socket.connect(dispatcher_endpoint)

    # result = Fmi2Return()
    # result.Fmi2ExtHandshakeReturn.SetInParent()

    state = Fmi2ExtHandshakeReturn().SerializeToString()
    socket.send(state)

    command = Fmi2Command()

    while True:

        msg = socket.recv()
        command.ParseFromString(msg)

        group = command.WhichOneof("command")

        data = getattr(command, command.WhichOneof("command"))
        if group == "Fmi2SetupExperiment":
            result = Fmi2StatusReturn()
            result.status = slave.fmi2SetupExperiment(
                data.start_time, data.stop_time, data.tolerance
            )

        elif group == "Fmi2DoStep":
            result = Fmi2StatusReturn()
            result.status = slave.fmi2DoStep(
                data.current_time, data.step_size, data.no_step_prior
            )
        elif group == "Fmi2EnterInitializationMode":
            result = Fmi2StatusReturn()
            result.status = slave.fmi2EnterInitializationMode()
        elif group == "Fmi2ExitInitializationMode":
            result = Fmi2StatusReturn()
            result.status = slave.fmi2ExitInitializationMode()
        elif group == "Fmi2ExtSerializeSlave":
            result = Fmi2ExtSerializeSlaveReturn()
            (result.status, result.state) = slave.fmi2ExtSerialize()
        elif group == "Fmi2ExtDeserializeSlave":
            state = command.Fmi2ExtDeserializeSlave.state
            result.Fmi2StatusReturn.status = slave.fmi2ExtDeserialize(state)
        elif group == "Fmi2GetReal":
            status, values = slave.fmi2GetReal(command.Fmi2GetReal.references)
            result.Fmi2GetRealReturn.status = status
            result.Fmi2GetRealReturn.values[:] = values
        elif group == "Fmi2GetInteger":
            status, values = slave.fmi2GetInteger(command.Fmi2GetInteger.references)
            result.Fmi2GetIntegerReturn.status = status
            result.Fmi2GetIntegerReturn.values[:] = values
        elif group == "Fmi2GetBoolean":
            status, values = slave.fmi2GetBoolean(command.Fmi2GetBoolean.references)
            result.Fmi2GetBooleanReturn.status = status
            result.Fmi2GetBooleanReturn.values[:] = values
        elif group == "Fmi2GetString":
            status, values = slave.fmi2GetString(command.Fmi2GetString.references)
            result.Fmi2GetStringReturn.status = status
            result.Fmi2GetStringReturn.values[:] = values
        elif group == "Fmi2SetReal":
            status = slave.fmi2SetReal(
                command.Fmi2SetReal.references, command.Fmi2SetReal.values
            )
            result.Fmi2StatusReturn.status = status
        elif group == "Fmi2SetInteger":
            status = slave.fmi2SetInteger(
                command.Fmi2SetInteger.references, command.Fmi2SetInteger.values
            )
            result.Fmi2StatusReturn.status = status
        elif group == "Fmi2SetBoolean":
            status = slave.fmi2SetBoolean(
                command.Fmi2SetBoolean.references, command.Fmi2SetBoolean.values
            )
            result.Fmi2StatusReturn.status = status
        elif group == "Fmi2SetString":
            status = slave.fmi2SetString(
                command.Fmi2SetString.references, command.Fmi2SetString.values
            )
            result.Fmi2StatusReturn.status = status
        elif group == "Fmi2Terminate":
            result.Fmi2StatusReturn.status = slave.fmi2Terminate()
        elif group == "Fmi2Reset":
            result.Fmi2StatusReturn.status = slave.fmi2Reset()
        elif group == "Fmi2FreeInstance":
            result.Fmi2FreeInstanceReturn.SetInParent()
            pass
        else:
            logger.error(f"unrecognized command '{group}' received, shutting down")
            sys.exit(-1)

        state = result.SerializeToString()
        socket.send(state)

        if group == "Fmi2FreeInstance":
            logger.info(f"Fmi2FreeInstance received, shutting down")
            sys.exit(0)

File: tool/unifmu/resources/backends/csharp/backend.cs

using System.IO;
using System;
using Fmi2Proto;
using System.Collections.Generic;
using NetMQ.Sockets;
using Google.Protobuf;
using NetMQ;


namespace Launch
{
  class Program
  {


    public static void Main(string[] args)
    {
      var references_to_attr = new Dictionary<uint, string>();
      var model = new Model();

      string dispatcher_endpoint = System.Environment.GetEnvironmentVariable("UNIFMU_DISPATCHER_ENDPOINT");
      if (dispatcher_endpoint == null)
      {
        Console.Error.WriteLine("Environment variable 'UNIFMU_DISPATCHER_ENDPOINT' is not set in the current enviornment.");
        Environment.Exit(-1);
      }


      var socket = new RequestSocket();
      socket.Connect(dispatcher_endpoint);

      var result = new Fmi2Return();
      result.Fmi2StatusReturn = new Fmi2StatusReturn();
      result.Fmi2GetRealReturn = new Fmi2GetRealReturn();
      result.Fmi2GetIntegerReturn = new Fmi2GetIntegerReturn();
      result.Fmi2GetBooleanReturn = new Fmi2GetBooleanReturn();
      result.Fmi2GetStringReturn = new Fmi2GetStringReturn();
      result.Fmi2GetStringReturn = new Fmi2GetStringReturn();
      result.Fmi2ExtSerializeSlaveReturn = new Fmi2ExtSerializeSlaveReturn();
      result.Fmi2ExtHandshakeReturn = new Fmi2ExtHandshakeReturn();

      socket.SendFrame(result.ToByteArray(), false);
      Fmi2Command command;

      while (true)
      {
        command = Fmi2Command.Parser.ParseFrom(socket.ReceiveFrameBytes());

        switch (command.CommandCase)
        {
          case Fmi2Command.CommandOneofCase.Fmi2SetupExperiment:
            {
              result.Fmi2StatusReturn = new Fmi2StatusReturn();

              var status = model.Fmi2SetupExperiment(
                   command.Fmi2SetupExperiment.StartTime,
                   command.Fmi2SetupExperiment.HasStopTime ? command.Fmi2SetupExperiment.StopTime : null,
                   command.Fmi2SetupExperiment.HasTolerance ? command.Fmi2SetupExperiment.Tolerance : null
               );

              result.Fmi2StatusReturn.Status = status;
            }
            break;

          case Fmi2Command.CommandOneofCase.Fmi2EnterInitializationMode:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2EnterInitializationMode();
            break;

          case Fmi2Command.CommandOneofCase.Fmi2ExitInitializationMode:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2ExitInitializationMode();
            break;

          case Fmi2Command.CommandOneofCase.Fmi2DoStep:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2DoStep(command.Fmi2DoStep.CurrentTime, command.Fmi2DoStep.StepSize, command.Fmi2DoStep.NoStepPrior);
            break;

          case Fmi2Command.CommandOneofCase.Fmi2SetReal:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.FmiSetReal(command.Fmi2SetReal.References, command.Fmi2SetReal.Values);
            break;


          case Fmi2Command.CommandOneofCase.Fmi2SetInteger:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2SetInteger(
                command.Fmi2SetInteger.References,
                command.Fmi2SetInteger.Values);
            break;

          case Fmi2Command.CommandOneofCase.Fmi2SetBoolean:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2SetBoolean(
                command.Fmi2SetBoolean.References,
                command.Fmi2SetBoolean.Values
            );
            break;


          case Fmi2Command.CommandOneofCase.Fmi2SetString:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2SetString(command.Fmi2SetString.References, command.Fmi2SetString.Values);
            break;

          case Fmi2Command.CommandOneofCase.Fmi2GetReal:
            {
              result.Fmi2GetRealReturn = new Fmi2GetRealReturn();
              (var status, var values) = model.Fmi2GetReal(command.Fmi2GetReal.References);
              result.Fmi2GetRealReturn.Values.AddRange(values);
              result.Fmi2GetRealReturn.Status = status;
            }
            break;

          case Fmi2Command.CommandOneofCase.Fmi2GetInteger:
            {
              result.Fmi2GetIntegerReturn = new Fmi2GetIntegerReturn();
              (var status, var values) = model.Fmi2GetInteger(command.Fmi2GetInteger.References);
              result.Fmi2GetIntegerReturn.Values.AddRange(values);
              result.Fmi2GetIntegerReturn.Status = status;
            }
            break;

          case Fmi2Command.CommandOneofCase.Fmi2GetBoolean:
            {
              result.Fmi2GetBooleanReturn = new Fmi2GetBooleanReturn();
              (var status, var values) = model.Fmi2GetBoolean(command.Fmi2GetBoolean.References);
              result.Fmi2GetBooleanReturn.Values.AddRange(values);
              result.Fmi2GetBooleanReturn.Status = status;
            }
            break;

          case Fmi2Command.CommandOneofCase.Fmi2GetString:
            {
              result.Fmi2GetStringReturn = new Fmi2GetStringReturn();
              (var status, var values) = model.Fmi2GetString(command.Fmi2GetString.References);
              result.Fmi2GetStringReturn.Values.AddRange(values);
              result.Fmi2GetStringReturn.Status = status;
            }
            break;

          case Fmi2Command.CommandOneofCase.Fmi2CancelStep:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2CancelStep();
            break;

          case Fmi2Command.CommandOneofCase.Fmi2Reset:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2Reset();
            break;

          case Fmi2Command.CommandOneofCase.Fmi2Terminate:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2Terminate();
            break;

          case Fmi2Command.CommandOneofCase.Fmi2ExtSerializeSlave:
            {
              result.Fmi2ExtSerializeSlaveReturn = new Fmi2ExtSerializeSlaveReturn();
              (var status, var state) = model.Fmi2ExtSerialize();
              result.Fmi2ExtSerializeSlaveReturn.Status = status;
              result.Fmi2ExtSerializeSlaveReturn.State = ByteString.CopyFrom(state);

            }
            break;
          case Fmi2Command.CommandOneofCase.Fmi2ExtDeserializeSlave:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = model.Fmi2ExtDeserialize(command.Fmi2ExtDeserializeSlave.State.ToByteArray());
            break;

          case Fmi2Command.CommandOneofCase.Fmi2FreeInstance:
            result.Fmi2StatusReturn = new Fmi2StatusReturn();
            result.Fmi2StatusReturn.Status = Fmi2Status.Ok;
            break;

          default:
            Console.Error.WriteLine("unrecognized command {0}, shutting down", command.CommandCase);
            Environment.Exit(-1);
            break;
        }


        socket.SendFrame(result.ToByteArray(), false);
        Console.WriteLine("returning result {0}", result);

        if (command.CommandCase == Fmi2Command.CommandOneofCase.Fmi2FreeInstance)
        {
          Environment.Exit(0);
        }

      }

    }
  }
}